src/App.jsx

```diff
@@ -1,33 +1,153 @@
+        if (!resp.ok) throw new Error(resp.statusText || "Failed to fetch todos");
+
+        const { records } = await resp.json();
+
+        const fetchedTodos = records.map((record) => {
+          const todo = { id: record.id, ...record.fields };
+          if (!todo.isCompleted) todo.isCompleted = false;
+          return todo;
+        });
+
+        setTodoList(fetchedTodos);
+      } catch (error) {
+        setErrorMessage(error.message);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  // ✅ Add new todo
+  const addTodo = async (title) => {
+    const newTodo = { title, isCompleted: false };
+
+    const payload = {
+      records: [
+        {
+          fields: {
+            title: newTodo.title,
+            isCompleted: newTodo.isCompleted,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: "POST",
+      headers: {
+        Authorization: token,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(payload),
     };
-    setTodoList([...todolist, newTodo]);
+
+    try {
+      setIsSaving(true);
+
+      const resp = await fetch(url, options);
+      if (!resp.ok) throw new Error(resp.statusText || "Failed to save todo");
+
+      const { records } = await resp.json();
+
+      const savedTodo = {
+        id: records[0].id,
+        ...records[0].fields,
+      };
+
+      if (!savedTodo.isCompleted) savedTodo.isCompleted = false;
+
+      setTodoList([...todolist, savedTodo]);
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(error.message);
+    } finally {
+      setIsSaving(false);
+    }
   };
 
-  // NUEVO: función para actualizar un todo
-  const updateTodo = (editedTodo) => {
+  // ✅ Update todo with optimistic approach
+  const updateTodo = async (editedTodo) => {
+    const originalTodo = todolist.find((todo) => todo.id === editedTodo.id);
+
+    // Optimistically update UI
     const updatedTodos = todolist.map((todo) =>
       todo.id === editedTodo.id ? { ...editedTodo } : todo
     );
     setTodoList(updatedTodos);
+
+    const payload = {
+      records: [
+        {
+          id: editedTodo.id,
+          fields: {
+            title: editedTodo.title,
+            isCompleted: editedTodo.isCompleted,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: "PATCH",
+      headers: {
+        Authorization: token,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+      if (!resp.ok) throw new Error(resp.statusText || "Failed to update todo");
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(`${error.message}. Reverting todo...`);
+
+      // Rollback if failure
+      const revertedTodos = todolist.map((todo) =>
+        todo.id === originalTodo.id ? originalTodo : todo
```

Comment:
Mutation hides changes. So it's good practice to return a copy of an object that is expected to change. e.g. `originalTodo`

Reply:
Be sure you apply this change in all other instances of this code block.